File: pr1-java/src/main/java/repository/database/TicketDbRepository.java
package repository.database;

import domain.Ticket;
import domain.exceptions.DuplicateException;
import domain.exceptions.NotFoundException;
import repository.TicketRepository;
import utils.Configuration;

import java.sql.Connection;
import java.sql.PreparedStatement;
import java.sql.ResultSet;
import java.sql.SQLException;
import java.util.ArrayList;
import java.util.List;

public class TicketDbRepository implements TicketRepository {
    public TicketDbRepository() {
        Configuration.logger.info("Initializing TicketDbRepository with properties: {} ", Configuration.properties);
    }

    @Override
    public Iterable<Ticket> getAll() {
        Configuration.logger.traceEntry();

        Connection connection = DbUtils.getConnection();
        List<Ticket> tickets = new ArrayList<>();
        try (ResultSet result = connection.prepareStatement("select * from tickets inner join games on games.gameId = tickets.forGameId;").executeQuery()) {
            while (result.next()) {
                Integer id = result.getInt("ticketId");
                String clientName = result.getString("clientName");
                Ticket ticket = new Ticket(id, GameDbRepository.getGameFromResultSet(result), clientName);
                tickets.add(ticket);
            }
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
        }

        Configuration.logger.traceExit(tickets);
        return tickets;
    }

    @Override
    public Ticket getOne(Integer id) throws NotFoundException {
        Configuration.logger.traceEntry();

        Connection connection = DbUtils.getConnection();
        Ticket ticket = null;
        try (ResultSet result = connection.prepareStatement("select * from tickets inner join games on games.gameId = tickets.forGameId where ticketId = " + id + ";").executeQuery()) {
            String clientName = result.getString("clientName");
            ticket = new Ticket(id, GameDbRepository.getGameFromResultSet(result), clientName);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
            throw new NotFoundException();
        }

        Configuration.logger.traceExit(ticket);
        return ticket;
    }

    @Override
    public Ticket add(Ticket ticket) {
        Configuration.logger.traceEntry("Saving ticket {} ", ticket);

        Connection connection = DbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("insert into tickets(forGameId, clientName) " + "values(?,?);")) {
            preparedStatement.setInt(1, ticket.getForGame().getId());
            preparedStatement.setString(2, ticket.getClientName());

            int result = preparedStatement.executeUpdate();
            Configuration.logger.trace("Saved {} instances", result);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
        }

        try (ResultSet result = connection.prepareStatement("select max(ticketId) as ticketId from tickets;").executeQuery()) {
            ticket.setId(result.getInt("ticketId"));
            Configuration.logger.trace("Got id for {}", ticket);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
        }

        Configuration.logger.traceExit(ticket);
        return ticket;
    }

    @Override
    public Ticket remove(Integer id) throws NotFoundException {
        Configuration.logger.traceEntry("Removing ticket with id {} ", id);

        Ticket ticket = getOne(id);
        Connection connection = DbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("delete from tickets where ticketId = " + id + ";")) {
            int result = preparedStatement.executeUpdate();
            Configuration.logger.trace("Deleted {} instances", result);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
            throw new NotFoundException();
        }

        Configuration.logger.traceExit(ticket);
        return ticket;
    }

    @Override
    public Ticket modify(Integer id, Ticket newTicket) throws NotFoundException {
        Configuration.logger.traceEntry("Modifying ticket with id {} ", id);

        Ticket oldTicket = getOne(id);
        Connection connection = DbUtils.getConnection();
        try (PreparedStatement preparedStatement = connection.prepareStatement("update tickets set " +
                "forGameId = " + newTicket.getForGame().getId() +
                ", clientName = '" + newTicket.getClientName() +
                "' where ticketId = " + id + ";")) {
            int result = preparedStatement.executeUpdate();
            Configuration.logger.trace("Modified {} instances", result);
        } catch (SQLException exception) {
            Configuration.logger.error(exception);
        }

        Configuration.logger.traceExit(oldTicket);
        return oldTicket;
    }

    @Override
    public String toString() {
        StringBuilder string = new StringBuilder();
        for (Ticket ticket : getAll())
            string.append(ticket).append("\n");
        return string.toString();
    }
}

File: pr1-cs/Persistence/Database/TicketDbRepository.cs
using System.Collections.Generic;
using System.Linq;
using Model;
using Model.Exceptions;
using Mono.Data.Sqlite;
using DuplicateException = Model.Exceptions.Thrift.DuplicateException;
using NotFoundException = Model.Exceptions.Thrift.NotFoundException;

namespace Persistence.Database
{
    public class TicketDbRepository : ITicketRepository
    {
        public IEnumerable<Ticket> GetAll()
        {
            var tickets = new List<Ticket>();
            var connection = DbUtils.Connection;
            using var command = connection.CreateCommand();
            command.CommandText = "select * from tickets inner join games on games.gameId = tickets.forGameId;";
            using var dataReader = command.ExecuteReader();
            while (dataReader.Read())
            {
                var ticketId = dataReader.GetInt32(dataReader.GetOrdinal("ticketId"));
                var clientName = dataReader.GetString(dataReader.GetOrdinal("clientName"));

                var ticket = new Ticket(ticketId, GameDbRepository.GetGameFromDataReader(dataReader),
                    clientName);
                tickets.Add(ticket);
            }

            return tickets;
        }

        public Ticket GetOne(int id)
        {
            Ticket ticket;
            var connection = DbUtils.Connection;
            using var command = connection.CreateCommand();
            command.CommandText =
                "select * from tickets inner join games on games.gameId = tickets.forGameId where ticketId = " +
                id + ";";
            using var dataReader = command.ExecuteReader();
            if (dataReader.Read())
            {
                var ticketId = dataReader.GetInt32(dataReader.GetOrdinal("ticketId"));
                var clientName = dataReader.GetString(dataReader.GetOrdinal("clientName"));

                ticket = new Ticket(ticketId, GameDbRepository.GetGameFromDataReader(dataReader), clientName);
            }
            else
            {
                throw new NotFoundException();
            }
            return ticket;
        }

        public Ticket Add(Ticket ticket)
        {
            var connection = DbUtils.Connection;
            using var command = connection.CreateCommand();
            command.CommandText =
                "insert into tickets(forGameId, clientName) values(@forGameId, @clientName);";

            var dataParameter = command.CreateParameter();
            dataParameter.ParameterName = "@forGameId";
            dataParameter.Value = ticket.ForGame.Id;
            command.Parameters.Add(dataParameter);

            dataParameter = command.CreateParameter();
            dataParameter.ParameterName = "@clientName";
            dataParameter.Value = ticket.ClientName;
            command.Parameters.Add(dataParameter);

            try
            {
                command.ExecuteNonQuery();
            }
            catch (SqliteException exception)
            {
                throw new DuplicateException();
            }

            command.CommandText = "select max(ticketId) as ticketId from tickets;";
            using var dataReader = command.ExecuteReader();
            dataReader.Read();
            var ticketId = dataReader.GetInt32(dataReader.GetOrdinal("ticketId"));
            ticket.Id = ticketId;
            return ticket;
        }

        public Ticket Remove(int id)
        {
            var ticket = GetOne(id);
            var connection = DbUtils.Connection;
            using var command = connection.CreateCommand();
            command.CommandText = "delete from tickets where ticketId = " + id + ";";
            var result = command.ExecuteNonQuery();
            if (result == 0)
                throw new NotFoundException();
            return ticket;
        }

        public Ticket Modify(int id, Ticket newTicket)
        {
            var oldTicket = GetOne(id);
            var connection = DbUtils.Connection;
            using var command = connection.CreateCommand();
            command.CommandText = "update tickets set " +
                                  "forGameId = " + newTicket.ForGame.Id +
                                  ", clientName = '" + newTicket.ClientName +
                                  "' where ticketId = " + id + ";";
            var result = command.ExecuteNonQuery();
            if (result == 0)
                throw new NotFoundException();
            return oldTicket;
        }

        public override string ToString()
        {
            return GetAll().Aggregate("", (current, ticket) => current + (ticket + "\n"));
        }
    }
}